opm/simulators/wells/WellGroupHelpers.hpp

```diff
@@ -0,0 +1,211 @@
+
+
+    namespace wellGroupHelpers
+    {
+
+    inline void setCmodeGroup(const Group2& group, const Schedule& schedule, const SummaryState& summaryState, const int reportStepIdx, WellStateFullyImplicitBlackoil& wellState) {
+
+        for (const std::string& groupName : group.groups()) {
+            setCmodeGroup( schedule.getGroup2(groupName, reportStepIdx), schedule, summaryState, reportStepIdx, wellState);
+        }
+
+        if (group.isInjectionGroup()) {
+            const auto controls = group.injectionControls(summaryState);
+            wellState.setCurrentInjectionGroupControl(group.name(), GroupInjection::ControlEnum2String(controls.cmode));
+        }
+        if (group.isProductionGroup()) {
+            const auto controls = group.productionControls(summaryState);
+            wellState.setCurrentProductionGroupControl(group.name(), GroupProduction::ControlEnum2String(controls.cmode));
+        }
+    }
+
+
+    inline void accumulateGroupEfficiencyFactor(const Group2& group, const Schedule& schedule, const int reportStepIdx, double& factor) {
+        factor *= group.getGroupEfficiencyFactor();
+        if (group.parent() != "FIELD")
+            accumulateGroupEfficiencyFactor(schedule.getGroup2(group.parent(), reportStepIdx), schedule, reportStepIdx, factor);
+    }
+
+
+    inline void sumWellRates(const Group2& group, const Schedule& schedule, const WellStateFullyImplicitBlackoil& wellState, const int reportStepIdx, const int phasePos, const bool injector, double& rate) {
+
+        for (const std::string& groupName : group.groups()) {
+            const Group2& groupTmp = schedule.getGroup2(groupName, reportStepIdx);
+            sumWellRates(groupTmp, schedule, wellState, reportStepIdx, phasePos, injector, rate);
+            rate *= groupTmp.getGroupEfficiencyFactor();
+        }
+        const auto& end = wellState.wellMap().end();
+        for (const std::string& wellName : group.wells()) {
+            const auto& it = wellState.wellMap().find( wellName );
+            if (it == end)  // the well is not found
+                continue;
+
+            int well_index = it->second[0];
+
+            const auto& wellEcl = schedule.getWell2(wellName, reportStepIdx);
+            //only count producers or injectors
+            if ( (wellEcl.isProducer() && injector) ||  (wellEcl.isInjector() && !injector))
+                continue;
+
+            double factor = wellEcl.getEfficiencyFactor();
+            const auto wellrate_index = well_index * wellState.numPhases();
+            if (injector)
+                rate += factor * wellState.wellRates()[ wellrate_index + phasePos];
+            else
+                rate -= factor * wellState.wellRates()[ wellrate_index + phasePos];
+        }
+    }
+
+    inline void sumWellResRates(const Group2& group, const Schedule& schedule, const WellStateFullyImplicitBlackoil& wellState, const int reportStepIdx, const int phasePos, const bool injector, double& rate) {
```

Comment:
These functions should be refactored to not repeat all of it.

Reply:
These functions are temporary since they have to be rewritten when wellState is refactored. As I see it there is no need to spend time on refactoring it at the time.